crates/ruff/src/rules/ruff/rules/unreachable.rs

```diff
@@ -0,0 +1,964 @@
+    /// constants of [`BasicBlock`].
+    blocks: Vec<BasicBlock<'stmt>>,
+}
+
+impl<'stmt> From<&'stmt [Stmt]> for BasicBlocks<'stmt> {
+    /// # Notes
+    ///
+    /// This assumes that `stmts` is a function body.
+    fn from(stmts: &'stmt [Stmt]) -> BasicBlocks<'stmt> {
+        let mut blocks = Vec::with_capacity(stmts.len());
+
+        create_blocks(&mut blocks, stmts, None);
+
+        if blocks.is_empty() {
+            blocks.push(BasicBlock::EMPTY);
+        }
+
+        BasicBlocks { blocks }
+    }
+}
+
+/// Basic code block, sequence of statements unconditionally executed
+/// "together".
+#[derive(Debug, PartialEq)]
+struct BasicBlock<'stmt> {
+    stmts: &'stmt [Stmt],
+    next: NextBlock<'stmt>,
+}
+
+/// Edge between basic blocks (in the control-flow graph).
+#[derive(Debug, PartialEq)]
+enum NextBlock<'stmt> {
+    /// Always continue with a block.
+    Always(BlockIndex),
+    /// Condition jump.
+    If {
+        /// Condition that needs to be evaluated to jump to the `next` or
+        /// `orelse` block.
+        condition: Condition<'stmt>,
+        /// Next block if `condition` is true.
+        next: BlockIndex,
+        /// Next block if `condition` is false.
+        orelse: BlockIndex,
+    },
+    /// The end.
+    Terminate,
+}
+
+/// Condition used to determine to take the `next` or `orelse` branch in
+/// [`NextBlock::If`].
+#[derive(Clone, Debug, PartialEq)]
+enum Condition<'stmt> {
+    /// Conditional statement, this should evaluate to a boolean, for e.g. `if`
+    /// or `while`.
+    Test(&'stmt Expr),
+    /// Iterator for `for` statements, e.g. for `i in range(10)` this will be
+    /// `range(10)`.
+    Iterator(&'stmt Expr),
+    Match {
+        /// `match $subject`.
+        subject: &'stmt Expr,
+        /// `case $case`, include pattern, guard, etc.
+        case: &'stmt MatchCase,
+    },
+}
+
+impl<'stmt> Ranged for Condition<'stmt> {
+    fn range(&self) -> TextRange {
+        match self {
+            Condition::Test(expr) | Condition::Iterator(expr) => expr.range(),
+            // The case of the match statement, without the body.
+            Condition::Match { subject: _, case } => TextRange::new(
+                case.start(),
+                case.guard
+                    .as_ref()
+                    .map_or(case.pattern.end(), |guard| guard.end()),
+            ),
+        }
+    }
+}
+
+/// Index into [`BasicBlocks::blocks`].
+type BlockIndex = usize;
+
+impl<'stmt> BasicBlock<'stmt> {
+    /// A sentinel block indicating an empty termination block.
+    const EMPTY: BasicBlock<'static> = BasicBlock {
+        stmts: &[],
+        next: NextBlock::Terminate,
+    };
+
+    /// A sentinel block indicating an exception was raised.
+    const EXCEPTION: BasicBlock<'static> = BasicBlock {
+        stmts: &[Stmt::Return(StmtReturn {
+            range: TextRange::new(TextSize::new(0), TextSize::new(0)),
+            value: None,
+        })],
```

Comment:
Is there a risk that creating a sentinel return statement drips over any analysis or diagnostic rendering?

Edit: Okay. I guess you already handle this by `is_sentinel`

Reply:
Sentinel blocks are printed in the graph, which looks a little odd at times, but indeed in the diagnostics they shouldn't appear.